packages/storage/amplify_storage_s3_dart/lib/src/storage_s3_service/service/task/s3_upload_task.dart

```diff
@@ -535,11 +538,15 @@ class S3UploadTask {
     // create a new batching Completer to start the current batching
     _uploadPartBatchingCompleter = Completer();
 
-    final numToBatch = min(
-      _maxNumParallelTasks - _numOfOngoingSubtasks,
-      _expectedNumOfSubtasks -
-          (_numOfCompletedSubtasks + _numOfOngoingSubtasks),
-    );
+    // If the AWSFile is backed by a stream, then upload chunks one by one
+    // otherwise maximum 4 parts of the file can be read and uploaded in parallel
```

Comment:
Why can't a stream be uploaded in parallel?

Reply:
Stream cannot be read as chunks in parallel with random start and end positions, e.g. `ChunkedStreamReader` `readChunk` returns a Future of collected bytes that needs to be await'ed and doesn't allow calls in parallel. 

We surely can read a few chunks in advance caching the bytes in memory and then start uploads in parallel, but caching bytes in memory could lead to memory leak.

Although I could use additional optimization here, e.g. checking part size, if it's still 5MB, that we can cache 20MB bytes in memory to allow uploads in parallel.